FixMath2.h

```diff
@@ -0,0 +1,845 @@
+    constexpr byte new_NF = MAX(NF, _NF);
+    typedef typename IntegerType< ((new_NI+new_NF-1)>>3)+1>::unsigned_type return_type;
+    UFixMath2<new_NI,new_NF> left(*this);
+    UFixMath2<new_NI,new_NF> right(op);
+
+    return_type tt = return_type(left.asRaw()) + right.asRaw();
+    return UFixMath2<new_NI,new_NF>(tt,true);
+  }
+
+  /** Addition with another type. Unsafe
+@param op The number to be added.
+@return The result of the addition as a UFixMath2.
+  */
+  template<typename T>
+  UFixMath2<NI,NF> operator+ (const T op) const
+  {
+    return UFixMath2<NI,NF>(internal_value+((internal_type)op<<NF),true);
+  }
+
+
+  //////// SUBSTRACTION OVERLOADS
+
+      /** Subtraction with another UFixMath2. Safe.
+@param op The UFixMath2 to be subtracted.
+@return The result of the subtraction as a UFixMath2.
+  */
+  template<byte _NI, byte _NF> // We do not have the +1 after MAX(NI, _NI) because the substraction between two UFix should fit in the bigger of the two.
+  UFixMath2<MAX(NI,_NI),MAX(NF,_NF)> operator- (const UFixMath2<_NI,_NF>& op) const
+  {
+    constexpr byte new_NI = MAX(NI, _NI);
+    constexpr byte new_NF = MAX(NF, _NF);
+    typedef typename IntegerType< ((new_NI+new_NF-1)>>3)+1>::unsigned_type return_type;
+    UFixMath2<new_NI,new_NF> left(*this);
+    UFixMath2<new_NI,new_NF> right(op);
+
+    return_type tt = return_type(left.asRaw()) - right.asRaw();
+    return UFixMath2<new_NI,new_NF>(tt,true);
+  }
+
+  
+    /** Subtraction with another type. Unsafe
+@param op The number to be subtracted.
+@return The result of the subtraction as a UFixMath2.
+  */
+  template<typename T>
+  UFixMath2<NI,NF> operator- (const T op) const
+  {
+    return UFixMath2<NI,NF>(internal_value-((internal_type)op<<NF),true);
+  }
+
+    /** Opposite of the number.
+@return The opposite numberas a SFixMath2.
+  */
+  SFixMath2<NI+1,NF> operator-() const
+  {
+    return SFixMath2<NI+1,NF>( -(typename IntegerType<((NI+NF-1+1)>>3)+1>::signed_type)(internal_value),true);
+  }
+
+  //////// MULTIPLICATION OVERLOADS
+  
+  /** Multiplication with another UFixMath2. Safe.
+@param op The UFixMath2 to be multiplied.
+@return The result of the multiplication as a UFixMath2.
+  */
+  template<byte _NI, byte _NF>
+  UFixMath2<NI+_NI,NF+_NF> operator* (const UFixMath2<_NI,_NF>& op) const
+  {
+    typedef typename IntegerType< ((NI+_NI+NF+_NF-1)>>3)+1>::unsigned_type return_type ;
+    return_type tt = return_type(internal_value)*op.asRaw();
+    return UFixMath2<NI+_NI,NF+_NF>(tt,true);
+  }
+
+  /** Multiplication with another type. Unsafe.
+@param op The number to be multiplied.
+@return The result of the multiplication as a UFixMath2.
+  */
+  template<typename T>
+  UFixMath2<NI,NF> operator* (const T op) const
+  {
+    return UFixMath2<NI,NF>(internal_value*op,true);
+  }
+
+
+  /** Right shift. This won't overflow but will not leverage on the bits freed by the shift.
+Better to use .sR<shift>() if possible instead.
```

Comment:
If I understand correctly, you implemented .sR() separately, after all. Why not as a member?

Reply:
`sL()` and `sR()` are implemented as members a bit further (l 263).

Reply:
Note that I realize now that .SR could be "just" moving the comma and not removing some bits on the right so that (pseudo-code):
```
UFixMath<8,8>.sR() -> UFixMath<6,10>
```
instead of 
```
UFixMath<8,8>.sR() -> UFixMath<6,8>
```
as it is now.

When I implemented it the first behavior seemed logical, but now I have the feeling that it does not align with the otherwise coherent paradigm of ensuring safety and keeping precision. Any opinion @tfry-git ?

Note that, if it is changed, the old behavior can easily be recover by enforcing the output type (conversion is the big plus of these classes, you can reduce precision easily with them):
```
UFixMath<8,8> a;
UFixMath<6,8> = a.sR(); // UFixMath<8,8>
```
instead of using `auto`:
```
auto b = a.sR(); // UFixMath<6,10>
```

I have the feeling a very comprehensive documentation will be useful for avoiding weird usages that goes outside the intended goal. Good thing that now the Arduino IDE does infer the type so the actual type represented by  `auto` is actually inferred in the IDE.

Reply:
Yes, I think aiming for coherency is indeed the way to go, so it really makes sense to extend it to sR()/sL().

But also, yes, good documentation and example(s) will be key.

A half-baked idea for brainstorming: Suppose we add a third template parameter "max bits", defaulting to 32 or 64. This will be carried along any operation (probably the smaller value, in case of ambiguity). Each operator will additionally get a static_assert to ensure that the total number of bits is still below this, with a hint to add a downcast, somewhere.

Reply:
That sounds reasonnable, I'll have to think about it. Note that there are already static_assert for preventing casting something bigger than 64. This automatically kick in if you try to do an operation which result exceed 64bits.